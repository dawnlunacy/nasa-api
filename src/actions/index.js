export const setNeos = neos => ({
  type: "SET_NEOS",
  neos
});

export const setTotalNeos = totalNeos => ({
  type: "SET_TOTAL_NEOS",
  totalNeos
});

export const setPrevWeek = prevWeekFetchUrl => ({
  type: "SET_PREV_WEEK",
  prevWeekFetchUrl
});

export const setNextWeek = nextWeekFetchUrl => ({
  type: "SET_NEXT_WEEK",
  nextWeekFetchUrl
});

export const updateLoading = bool => ({
  type: "UPDATE_LOADING",
  bool
})