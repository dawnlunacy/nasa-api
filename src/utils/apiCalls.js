export const fetchAPOD = async() => {
  const imageUrl = 'https://api.nasa.gov/planetary/apod?api_key=m98g3WmabopZXIZRCQ0HdHYrEwuHimuH8b8JjicA';
  const response = await fetch(imageUrl)
  if (response.ok) {
    const data = await response.json()
    return data.url;
  } else {
    throw Error("There was an error loading the Astronomy Picture Of the Day from NASA. Try again or look outside for incoming asteroids or comets.")
  }
}

export const fetchNEO = async(startDate, endDate) => {
  const defaultUrl = `https://api.nasa.gov/neo/rest/v1/feed?start_date=${startDate}&end_date=${endDate}&api_key=m98g3WmabopZXIZRCQ0HdHYrEwuHimuH8b8JjicAYOLO`;
  const response = await fetch(defaultUrl)
  if (response.ok) {
    const asteroidData = await response.json();
    return asteroidData;
  } else {
    throw Error("There was an error loading the NEAR EARTH OBJECTS from NASA. Try again or look outside for incoming asteroids or comets.");
  }
}

