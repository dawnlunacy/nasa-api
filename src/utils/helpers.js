export const findTodaysDate = () =>{
  let today = new Date();
  let dd = String(today.getDate()).padStart(2, '0');
  let mm = String(today.getMonth() + 1).padStart(2, '0');
  let yyyy = today.getFullYear();
  let todayFormat = yyyy + '/' + mm + '/' + dd;
  return todayFormat
}

export const formatTodaysDate = (stringDate) => {
  let formatDate = new Date(stringDate).toString().split(' ').slice(1, 4).join(' ')
  return formatDate
}

export const findDay = () => {
  let today = findTodaysDate();
  let weekday = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
  let day = weekday[new Date(today).getDay()]
  return day
}