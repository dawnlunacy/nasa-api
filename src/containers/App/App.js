
import React, { Component } from 'react';
import { connect } from 'react-redux'
import { Header } from '../Header/Header';
import { Nav } from '../Nav/Nav';
import AsteroidContainer from '../AsteroidContainer/AsteroidContainer';
import { fetchAPOD, fetchNEO } from '../../utils/apiCalls';
import { formatDateForFetch, findEndOfWeek, cleanNeoData } from '../../utils/helpers';
import { setNeos, setTotalNeos, setPrevWeek, setNextWeek, updateLoading, setCurrentNeoDate, setStartDate, setApod } from '../../actions';
import './App.css';

export class App extends Component {

  async componentDidMount() {
    this.getApod();
    this.saveNeosHelper();
  }

  startDateHelper = async (date) => {
    const { setStartDate } = this.props;
    const startOfWeek = formatDateForFetch(date)
    setStartDate(startOfWeek)
    this.saveNeosHelper(startOfWeek)
  }

  saveNeosHelper = async (startDate) => {
    const { setNeos, setTotalNeos, setPrevWeek, setNextWeek, isLoadingNeos} = this.props;
    isLoadingNeos(true)
    if (startDate === undefined) {
      startDate = formatDateForFetch()
    } 
    var endDate = findEndOfWeek(startDate);
    const neos = await fetchNEO(startDate, endDate);
    const cleanNeos = cleanNeoData(neos);
    setPrevWeek(neos.links.prev)
    setTotalNeos(neos.element_count)
    setNextWeek(neos.links.next)
    setNeos(cleanNeos)
    isLoadingNeos(false)
    
  }
  
  getApod = async() => {
    const { setApod } = this.props
    const backgroundImg = await fetchAPOD();

    const mainStyle = {
      backgroundImage:`url(${backgroundImg})`,
      backgroundRepeat: 'no-repeat',
      backgroundSize: '100% 100%',
    }
    console.log("apod", mainStyle)
    setApod(mainStyle);
  }

  displayDateSelectedNeos = (e) => {
    e.preventDefault();
    const { setCurrentNeoDate } = this.props;
    setCurrentNeoDate(e.target.id)
  }
  
  render() {
    const { loadingNeos } = this.props;
   
    return (
      <div className = "App">
        <Header />
        <Nav />
        {!loadingNeos && <AsteroidContainer 
          displayDateSelectedNeos={this.displayDateSelectedNeos}
          startDateHelper={this.startDateHelper}/> }
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  loadingNeos: state.loadingNeos
})

export const mapDispatchToProps = dispatch => ({
  setNeos: neos => dispatch (setNeos(neos)),
  setTotalNeos: totalNeos => dispatch (setTotalNeos(totalNeos)),
  setPrevWeek: prevWeekFetchUrl => dispatch(setPrevWeek(prevWeekFetchUrl)),
  setNextWeek: nextWeekFetchUrl => dispatch(setNextWeek(nextWeekFetchUrl)),
  isLoadingNeos: bool => dispatch(updateLoading(bool)),
  setCurrentNeoDate: date => dispatch(setCurrentNeoDate(date)),
  setStartDate: date => dispatch(setStartDate(date)),
  setApod: apod => dispatch(setApod(apod))
  })

export default connect(mapStateToProps, mapDispatchToProps)(App);