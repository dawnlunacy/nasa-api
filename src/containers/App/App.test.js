import React from 'react';
import { shallow } from 'enzyme';
import { App, mapStateToProps, mapDispatchToProps } from './App';
import { fetchAPOD, fetchNEO } from '../../utils/apiCalls';
import { mockUnfilteredApodResponse } from '../../utils/mockData';
import { mockNeoDataUnfiltered } from '../../utils/mockNeoDataUnfiltered';
// import { formatDateForFetch, findEndOfWeek, cleanNeoData } from '../../utils/helpers';
import { findDay, cleanNeoData } from '../../utils/helpers';
import { setNeos, setTotalNeos, setPrevWeek, setNextWeek, updateLoading, setCurrentNeoDate, setStartDate, setApod, handleError } from '../../actions';
import { mockCleanNeos } from '../../utils/mockData';


jest.mock('../../utils/apiCalls');
// jest.mock('../../utils/helpers');

describe('App', () => {
  let wrapper;
  const mockSetNeos = jest.fn();
  const mockSetTotalNeos = jest.fn();
  const mockSetPrevWeek = jest.fn();
  const mockSetNextWeek = jest.fn();
  const mockIsLoadingNeos = jest.fn();
  const mockSetCurrentNeoData = jest.fn();
  const mockSetStartDate = jest.fn();
  const mockSetApod = jest.fn();
  const mockFindEndOfWeek = jest.fn();
  const mockHandleError = jest.fn();
  

  fetchAPOD.mockImplementation(() => Promise.resolve(mockUnfilteredApodResponse));
  fetchNEO.mockImplementation(() => Promise.resolve(mockNeoDataUnfiltered));

  beforeEach(() => {
    wrapper = shallow(<App 
      loadingNeos={true}
      setNeos={mockSetNeos}
      setTotalNeps={mockSetTotalNeos}
      setPrevWeek={mockSetPrevWeek}
      setNextWeek={mockSetNextWeek}
      isLoadingNeos={mockIsLoadingNeos}
      setCurrentNeoDate={mockSetCurrentNeoData}
      setStartDate={mockSetStartDate}
      setApod={mockSetApod}
      handleError={mockHandleError}

      />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it.skip('should call the correct methods when saveNeosHelper', () => {
    const mockStartDate = '2019-11-03';
    const mockEndDate = '2019-11-10';
    wrapper.instance().findEndOfWeek = jest.fn()

     wrapper.instance().saveNeosHelper(mockStartDate);
    expect(wrapper.instance().findEndOfWeek).toHaveBeenCalled()
    expect(wrapper.instance().findEndOfWeek).toHaveBeenCalledWith(mockStartDate)
  })

  describe('mapStateToProps', () => {
    it('should return an object with loading and error info', () => {
      const mockLoading = false;
      const mockErrorMessage = "Error loading, please try again"
      const mockState = {
        loadingNeos: false,
        errorMessage: "Error loading, please try again"
      }

      const expected = {
        loadingNeos: mockLoading,
        errorMessage: mockErrorMessage
      }

      const mappedProps = mapStateToProps(mockState);
      
      expect(mappedProps).toEqual(expected)
    });
  });
  
  describe('mapDispatchToProps', () => {
    it('should call dispatch with a setNeos action when saveNeosHelper is called', () => {
      const mockDispatch = jest.fn();
      const mockResponse = mockCleanNeos;

      const actionToDispatch = setNeos(mockResponse);

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setNeos(mockResponse);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });

    it('should call dispatch with a setTotalNeos action when saveNeosHelper is called', () => {
      const mockDispatch = jest.fn();
      const mockResponse = 20

      const actionToDispatch = setTotalNeos(mockResponse);

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setTotalNeos(mockResponse);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });

    it('should call dispatch with a setPrevWeek action when saveNeosHelper is called', () => {
      const mockDispatch = jest.fn();
      const mockResponse = 'url'

      const actionToDispatch = setPrevWeek(mockResponse);

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setPrevWeek(mockResponse);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });

    it('should call dispatch with a setNextWeek action when saveNeosHelper is called', () => {
      const mockDispatch = jest.fn();
      const mockResponse = 'urlFuture'

      const actionToDispatch = setNextWeek(mockResponse);

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setNextWeek(mockResponse);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });

    it('should call dispatch with a updateLoading action when saveNeosHelper is called', () => {
      const mockDispatch = jest.fn();
      const mockResponse = false

      const actionToDispatch = updateLoading(mockResponse);

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.isLoadingNeos(mockResponse);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });

    it('should call dispatch with a setCurrentNeoDate action when displayDateSelectedNeos is called', () => {
      const mockDispatch = jest.fn();
      const mockResponse = '2019-11-08'

      const actionToDispatch = setCurrentNeoDate(mockResponse);

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setCurrentNeoDate(mockResponse);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });

    it('should call dispatch with a setStartDate action when startDateHelper is called', () => {
      const mockDispatch = jest.fn();
      const mockResponse = '2019-12-21'

      const actionToDispatch = setStartDate(mockResponse);

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setStartDate(mockResponse);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });

    it('should call dispatch with a setApod action when getApod is called', () => {
      const mockDispatch = jest.fn();
      const backgroundImg = 'https:wwww.APOD.com'
      const mockResponse =  {
        backgroundImage:`url(${backgroundImg})`,
        backgroundRepeat: 'no-repeat',
        backgroundSize: '100% 100%',
      }

      const actionToDispatch = setApod(mockResponse);

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setApod(mockResponse);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });

    it('should call dispatch with a handleError action when saveNeosHelper is called and a bad reponse is recieved from fetch', () => {
      const mockDispatch = jest.fn()
      const mockResponse =  "ERROR FETCHING"

      const actionToDispatch = handleError(mockResponse);

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.handleError(mockResponse);

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });
  });
});


