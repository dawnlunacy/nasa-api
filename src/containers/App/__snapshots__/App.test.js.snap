// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<div
  className="App"
>
  <Route
    exact={true}
    path="/(asteroids|)/"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/apod"
    render={[Function]}
  />
</div>
`;
